### PR TITLE
Add datastore fill pattern configuration

### PR DESCRIPTION
**Pull request by JamesClonk**
Add a configuration option to define filling pattern & sort order for datastores that the vsphere-cpi picks.

### CODE REVIEW

**Comment by JamesClonk** [comment]
Currently it always picks approx. the least used datastore with the most free space, but we'd like to be able to configure BOSH to choose the most used datastore, in order to fill them up efficiently.

**Comment by Lafunamor** [comment]
This would be really helpful

**Comment by nehagjain15** [comment]
@JamesClonk can you please add some unit test for storage_picker and datastore_picker using this new configuration
Once you do we can run test and merge it

**Comment by JamesClonk** [comment]
@nehagjain15 I've updated the PR accordingly. Added some unit-tests for the fill pattern configuration.
The default behaviour stays the same as before, but it now allows to configure the CPI to try and fill up the datastores more efficiently based on choosing either the most full or least full datastore to place disks.

**Comment by nehagjain15** [comment]
Thank you @JamesClonk.

**Comment by EleanorRigby** [comment]
@JamesClonk : Thanks for submitting the PR. 

I have some concerns regarding the core logic of datastore picker. I have added review comments in the code. Let's discuss them first before moving on to syntactic review (if needed).

**Comment by JamesClonk** [comment]
@EleanorRigby where can I see these comments?

File: src/vsphere_cpi/lib/cloud/vsphere/datastore_picker.rb [hunk 1]
```diff
@@ -44,12 +44,14 @@ def best_disk_placement(disks)
 
         if existing_ds_name =~ Regexp.new(disk.target_datastore_pattern) && datastores.keys.include?(existing_ds_name)
           datastores[existing_ds_name][:disks].push(disk)
+          break if datastore_fill_pattern != "" && datastore_fill_pattern != "none"
```
[the remark this answers is not in the document]
Comment (nehagjain15) [reply src/vsphere_cpi/lib/cloud/vsphere/datastore_picker.rb]: What happens with the other disks that we were looping on?

File: src/vsphere_cpi/lib/cloud/vsphere/datastore_picker.rb [hunk 2]
```diff
@@ -64,6 +66,8 @@ def best_disk_placement(disks)
             placement[:migration_size] += additional_required_space
             migration_size_accounted_for = true
           end
+
+          break if datastore_fill_pattern != "" && datastore_fill_pattern != "none"
```
[the remark this answers is not in the document]
Comment (nehagjain15) [reply src/vsphere_cpi/lib/cloud/vsphere/datastore_picker.rb]: I think this will violate the whole logic when few of the hosts are under maintenance mode

**Comment by EleanorRigby** [comment]
Closing this. As the entire architecture for cluster and datastore picker has changed.

**Pull request closed without merging**